File: app/src/main/java/ar/com/p39/localshare/sharer/ShareActivity.java
package ar.com.p39.localshare.sharer;

import android.content.Intent;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.provider.OpenableColumns;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.zxing.BarcodeFormat;
import com.google.zxing.EncodeHintType;
import com.google.zxing.MultiFormatWriter;
import com.google.zxing.WriterException;
import com.google.zxing.common.BitMatrix;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.EnumMap;
import java.util.Map;

import ar.com.p39.localshare.R;

public class ShareActivity extends AppCompatActivity {

    private TextView files;
    private QRImageView qr;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_share);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        files = (TextView) findViewById(R.id.files);
        qr = (QRImageView) findViewById(R.id.qr);

        Intent intent = getIntent();
        String action = intent.getAction();
        String type = intent.getType();

        // Figure out what to do based on the intent type
        if (Intent.ACTION_SEND.equals(action) && type != null) {
            if (type.startsWith("image/")) {
                handleSendImage(intent); // Handle single image being sent
            }
        } else if (Intent.ACTION_SEND_MULTIPLE.equals(action) && type != null) {
            if (type.startsWith("image/")) {
                handleSendMultipleImages(intent); // Handle multiple images being sent
            }
        } else {
            // Handle other intents, such as being started from the home screen
        }
    }


    void handleSendImage(Intent intent) {
        Uri imageUri = (Uri) intent.getParcelableExtra(Intent.EXTRA_STREAM);
        if (imageUri != null) {
            // Update UI to reflect image being shared
            files.append(imageUri.toString());
            files.append("\n");
            showUriData(imageUri);
        }
    }

    void handleSendMultipleImages(Intent intent) {
        ArrayList<Uri> imageUris = intent.getParcelableArrayListExtra(Intent.EXTRA_STREAM);
        if (imageUris != null) {
            // Update UI to reflect multiple images being shared
            for (Uri u : imageUris) {
                files.append(u.toString());
                files.append("\n");
                showUriData(u);
            }
        }
    }

    void showUriData(Uri uri) {
        Cursor returnCursor = getContentResolver().query(uri, null, null, null, null);
        /*
         * Get the column indexes of the data in the Cursor,
         * move to the first row in the Cursor, get the data,
         * and display it.
         */
        int nameIndex = returnCursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
        int sizeIndex = returnCursor.getColumnIndex(OpenableColumns.SIZE);
        returnCursor.moveToFirst();

        StringBuffer buffer = new StringBuffer();

        buffer.append(returnCursor.getString(nameIndex));
        buffer.append("\n");
        buffer.append(Long.toString(returnCursor.getLong(sizeIndex)));
        buffer.append("\n");
        returnCursor.close();

        InputStream input = null;
        try {
            input = getContentResolver().openInputStream(uri);
            if (input != null) {
//                byte[] bb = new byte[1000];
//                while (input.read(bb) != -1) {
//                }
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if (input != null) {
                try {
                    input.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }

        String data = buffer.toString();
        qr.setData(data);
        files.setText(data);
    }


}

File: app/src/main/java/ar/com/p39/localshare/sharer/ShareActivity.kt
package ar.com.p39.localshare.sharer

import android.content.Intent
import android.net.Uri
import android.net.wifi.WifiManager
import android.os.Bundle
import android.provider.OpenableColumns
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.view.View
import android.widget.TextView
import ar.com.p39.localshare.BuildConfig
import ar.com.p39.localshare.MyApplication
import ar.com.p39.localshare.R
import ar.com.p39.localshare.common.IpAddress
import ar.com.p39.localshare.common.network.WifiSSIDProvider
import ar.com.p39.localshare.common.ui.QRImageView
import ar.com.p39.localshare.sharer.models.FileShare
import ar.com.p39.localshare.sharer.presenters.SharePresenter
import ar.com.p39.localshare.sharer.views.ShareView
import butterknife.bindView
import com.crashlytics.android.answers.Answers
import com.crashlytics.android.answers.CustomEvent
import dagger.Module
import dagger.Provides
import dagger.Subcomponent
import java.io.IOException
import java.util.*
import javax.inject.Inject

class ShareActivity : AppCompatActivity(), ShareView {
    val qr: QRImageView by bindView(R.id.qr)
    val help: TextView by bindView(R.id.help)
    val helpWifi: TextView by bindView(R.id.help_wifi)

    @Inject
    lateinit var presenter: SharePresenter

    @Inject
    lateinit var bssidProvider: WifiSSIDProvider

    @Inject
    lateinit var wifiManager: WifiManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_share)

        MyApplication.graph.plus(ShareActivityModule(intent)).inject(this)

        val toolbar = findViewById(R.id.toolbar) as Toolbar?
        setSupportActionBar(toolbar)

        supportActionBar?.title = getString(R.string.title_activity_share)

        presenter.bindView(this)
        presenter.checkWifiStatus(bssidProvider.isConnected(), bssidProvider.getBSSID())

        helpWifi.text = getString(R.string.help_network, bssidProvider.getBSSID())
    }


    override fun showUriData(uri: Uri) {
        showUriData(listOf(uri))
    }

    override fun showUriData(uris: List<Uri>) {
        val files = ArrayList<FileShare>()

        for (uri in uris) {
            val returnCursor = contentResolver.query(uri, null, null, null, null)

            val nameIndex = returnCursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
            val sizeIndex = returnCursor.getColumnIndex(OpenableColumns.SIZE)
            returnCursor.moveToFirst()


            try {
                val share = FileShare(
                        returnCursor.getString(nameIndex),
                        returnCursor.getLong(sizeIndex),
                        contentResolver.getType(uri),
                        contentResolver.openInputStream(uri)
                )
                files.add(share)
            } catch (e: IOException) {
            } finally {
                returnCursor.close()
            }
        }
        shareFiles(files)
    }

    fun Double.format(digits: Int) = java.lang.String.format("%.${digits}f", this)

    private fun shareFiles(files: List<FileShare>) {
        val ip = wifiManager.IpAddress()

        if (ip != null) {
            presenter.startSharing(ip, files)

            qr.setData("http://$ip:8080/sharer")

            val totalSize: Double = files.map { it.size }.sum() / (1024 * 1024.0)

            trackFiles(files.size)
        } else {
            showWifiError()
        }
    }

    private fun trackFiles(size: Int) {
        // TODO : DI!
        if (!BuildConfig.DEBUG) {
            Answers.getInstance().logCustom(CustomEvent("Files Shared").putCustomAttribute("Count", size));
        }
    }

    override fun showWifiError() {
        Snackbar.make(help, R.string.connect_to_wifi    , Snackbar.LENGTH_INDEFINITE).show()
        help.visibility = View.GONE
        qr.visibility = View.GONE
    }

    override fun onStop() {
        super.onStop()
        presenter.stopSharing()
    }

    @Subcomponent(modules = arrayOf(ShareActivityModule::class))
    interface ShareActivityComponent {
        fun inject(shareActivity: ShareActivity);
    }

    @Module
    class ShareActivityModule(private var intent: Intent) {
        @Provides
        fun provideSharePresenter(): SharePresenter {
            return SharePresenter(intent)
        }
    }
}
